File: project/Boomerangβ0.4/Boomerang/Scene/StageSelect.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Input;

namespace Boomerang
{
    class StageSelect : IScene
    {
        private SceneType nextScene;
        private int cursorPosition;//カーソルの場所

        private GameManager gameManager;
        //使用する分だけ用意
        private Renderer renderer;
        private InputState inputState;
        private Sound sound;

        private bool isEnd;//シーン終了フラグ

        public StageSelect(GameManager gameManager)
        {
            this.gameManager = gameManager;
            inputState = gameManager.GetInputState();
            sound = gameManager.GetSound();
            renderer = gameManager.GetRenderer();
            //ランダムは未使用
        }
        public void Initialize()
        {
            cursorPosition = 60;
            nextScene = SceneType.Stage1;
            isEnd = false;
        }
        public void Update()
        {
            sound.PlayBGM("stageBGM");

            SceneCange();

            if (inputState.IsKeyDown(Keys.Left))
            {
                if (cursorPosition > 60)
                {
                    cursorPosition -= 240;
                    sound.PlaySE("inputSE");
                }
            }
            if (inputState.IsKeyDown(Keys.Right))
            {
                if (cursorPosition < 1020)
                {
                    cursorPosition += 240;
                    sound.PlaySE("inputSE");
                }
            }

            if (inputState.IsKeyDown(Keys.Space))
            {
                this.isEnd = true;
                sound.PlaySE("decideSE");
            }
        }
        public void Draw()
        {
            renderer.DrawTexture("stage", Vector2.Zero);
            renderer.DrawTexture("player_sd", new Vector2(cursorPosition, 550));
        }
        public void Shutdown()
        {
            sound.StopBGM();
        }
        public bool IsEnd()
        {
            return this.isEnd;
        }
        public SceneType Next()
        {
            return nextScene;//ゲームプレイへ
        }

        private void SceneCange()
        {
            if (cursorPosition == 100)
            {
                nextScene = SceneType.Stage1;
            }
            if (cursorPosition == 200)
            {
                nextScene = SceneType.Stage2;
            }
            if (cursorPosition == 300)
            {
                nextScene = SceneType.Stage3;
            }
            if (cursorPosition == 400)
            {
                nextScene = SceneType.Stage4;
            }
            if (cursorPosition == 500)
            {
                nextScene = SceneType.Stage5;
            }
        }
    }
}

File: project/Boomerang_0.3++/Boomerang_C++/OikakeCppPractice/src/Scenes/StageSelect/StageSelect.cpp
#include "StageSelect.h"
#include "Scenes/Base/Scene.h"

#include"World/World.h"
#include"Actor/Base/ActorGroup.h"
#include"Renderer/Renderer.h"
#include"Actor/Base/EventMessage.h"
#include"Application/Window/Window.h"
#include"Input/Input.h"

StageSelect::StageSelect(WorldPtr & world)
	: isEnd(false)
	, world(world)
	, renderer(Renderer::GetInstance())
{
	CursorPosition = 60;
}

StageSelect::~StageSelect()
{
}

void StageSelect::LoadAssets()
{
	renderer.LoadTexture(Assets::Texture::StageSelect, "stage.png");
	renderer.LoadTexture(Assets::Texture::Cursor, "player_sd.png");
}

void StageSelect::Initialize()
{
	shBGM = LoadSoundMem("asset/texture/stageBGM.mp3");
	shSE1 = LoadSoundMem("asset/texture/inputSE.wav");
	shSE2 = LoadSoundMem("asset/texture/decideSE.wav");
	PlaySoundMem(shBGM, DX_PLAYTYPE_LOOP);

	world->GetSceneShareValue().Initialize();
	world->Initialize();


	isEnd = false;

	//!ワールドのイベントメッセージを受け取る
	world->AddEventMessageListener([&](EventMessage message, void* param)
	{
		HandleMessage(message, param);
	});
}

void StageSelect::Update(float deltaTime)
{
	Input::GetInstance().Update();
	world->Update(deltaTime);

	if (Input::GetInstance().GetKeyBoard().IsDown(KEY_INPUT_RIGHT))
	{
		if (CursorPosition < 1020)
		{
			CursorPosition += 240;
			PlaySoundMem(shSE1, DX_PLAYTYPE_BACK);
		}
	}

	if (Input::GetInstance().GetKeyBoard().IsDown(KEY_INPUT_LEFT))
	{
		if (CursorPosition > 60)
		{
			CursorPosition -= 240;
			PlaySoundMem(shSE1, DX_PLAYTYPE_BACK);
		}
	}

	if (CursorPosition == 60)
	{
		nextScene =Scene::Stage1;
	}
	if (CursorPosition == 300)
	{
		nextScene = Scene::Stage2;
	}
	if (CursorPosition == 540)
	{
		nextScene = Scene::Stage3;
	}
	if (CursorPosition == 780)
	{
		nextScene = Scene::Stage4;
	}
	if (CursorPosition == 1020)
	{
		nextScene = Scene::Stage5;
	}

	if (Input::GetInstance().GetKeyBoard().IsDown(KEY_INPUT_SPACE)) 
	{
		PlaySoundMem(shSE2, DX_PLAYTYPE_NORMAL);
		isEnd = true;
	}
}

void StageSelect::Draw() const
{
	renderer.DrawTexture(Assets::Texture::StageSelect);
	renderer.DrawTexture(Assets::Texture::Cursor,Vector2(CursorPosition,550));
	world->Draw(renderer);
}

bool StageSelect::IsEnd() const
{
	return isEnd;
}

Scene StageSelect::Next() const
{
	return nextScene;//本当は選択したステージへ
}

void StageSelect::Finalize()
{
	StopSoundMem(shBGM);
	world->Finalize();
	renderer.Clear();
}

void StageSelect::HandleMessage(EventMessage message, void * param)
{
	if (message == EventMessage::GameEnd)
	{
		isEnd = true;
	}
}


